src/riak_kv_qry_compiler.erl

```diff
@@ -97,16 +97,22 @@ run_select(Select, Row, InitialState) ->
 
 %% @priv
 run_select2([], _, _, Acc) ->
```

Comment:
This can be done with body recursion, which will be faster, especially for the ColState traversal, as it avoids the double reverse on the lists in the results.

```` erlang
-spec run_select(SelectionSpec::[compiled_select()], Row::[any()]) ->
                        [any()].
run_select(Select, Row) ->
    %% the second argument is the state, if we're return row query results then
    %% there is no long running state
    run_select2(Select, Row, undefined).

run_select(Select, Row, InitialState) ->
    %% the second argument is the state, if we're return row query results then
    %% there is no long running state
    run_select2(Select, Row, InitialState).


run_select2([Fn | SelectTail], Row, RowState) ->
    prepend_select_columns(Fn(Row, RowState), SelectTail, Row, RowState);
run_select2([Fn | SelectTail], Row, [ColState1 | ColStateTail] ) ->
    prepend_select_columns(Fn(Row, ColState1), SelectTail, Row, ColStateTail).

%% Check if the select column is actually multiple columns, as returned by
%% SELECT * FROM, or the corner case SELECT *, my_col FROM. This cannot simply
%% be flattened because nulls are represented as empty lists.
prepend_select_columns([], Select, Row, RowState) ->
    [ [] | run_select2(Select, Row, RowState) ];
prepend_select_columns(Cols, Select, Row, RowState) ->
    Cols ++ run_select2(Select, Row, RowState).
````